### PR TITLE
IGNITE-20454 Sql. Add a query prefetch callback that is notified when data prefetching is complete.

### PR DESCRIPTION
**Pull request by xtern**
https://issues.apache.org/jira/browse/IGNITE-20454

### CODE REVIEW

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/AsyncSqlScriptCursorTest.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+    private static AsyncCursor<Integer> prepareTestCursor(PageLoadListener listener) {
+        AsyncCursor<Integer> data = new AsyncWrapper<>(IntStream.range(0, 10).boxed().collect(Collectors.toList()).iterator());
+        QueryTransactionWrapper tx = new QueryTransactionWrapper(NoOpTransaction.readOnly("TX"), true);
+        ResultSetMetadata rsMeta = new ResultSetMetadataImpl(Collections.emptyList());
+
+        return new AsyncSqlScriptCursor<>(SqlQueryType.QUERY, rsMeta, tx, data, listener);
+    }
+
+    private static boolean fetchAll(AsyncCursor<Integer> cursor) throws ExecutionException, InterruptedException, TimeoutException {
```
Comment (lowka) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/AsyncSqlScriptCursorTest.java:99]: I think we do not need the result of this method + this method can be rewritten into `do { .. } while(res.hasMore())`.
It also does not need timed variant of Future::get.

**Comment by lowka** [comment]
@xtern Maybe we can add a test case for `ExecutionServiceImplTest` that exercises that this listener is called?

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/AsyncSqlScriptCursorTest.java (hunk 1 above)
Comment (xtern) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/AsyncSqlScriptCursorTest.java]: Done

Comment (xtern) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/AsyncSqlScriptCursorTest.java]: Sorry, not done yet)

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/AsyncSqlCursorImplTest.java [hunk 2]
```diff
@@ -45,27 +58,26 @@
...
-        List<Integer> list = List.of(1, 2, 3);
 
         AsyncSqlCursorImpl<Integer> cursor = new AsyncSqlCursorImpl<>(SqlQueryType.QUERY, RESULT_SET_METADATA, txWrapper,
-                new AsyncWrapper<>(CompletableFuture.completedFuture(list.iterator()), Runnable::run));
+                new AsyncWrapper<>(CompletableFuture.completedFuture(ROWS.iterator()), Runnable::run));
 
         int requestRows = 2;
         BatchedResult<Integer> in1 = cursor.requestNextAsync(requestRows).join();
-        assertEquals(in1.items(), list.subList(0, requestRows));
+        assertEquals(ROWS.subList(0, requestRows), in1.items());
```
Comment (xtern) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/AsyncSqlCursorImplTest.java:74]: swapped args to (expected, result)

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/AsyncSqlCursorImplTest.java [hunk 3]
```diff
@@ -45,27 +58,26 @@
...
 
         assertFalse(((NoOpTransaction) txWrapper.unwrap()).commitFuture().isDone(),
                 "Implicit transaction should have not been committed because there is more data.");
 
-        BatchedResult<Integer> in2 = cursor.requestNextAsync(requestRows).join();
-        assertEquals(in2.items(), list.subList(requestRows, list.size()));
+        BatchedResult<Integer> in2 = cursor.requestNextAsync(ROWS.size()).join();
+        assertEquals(ROWS.subList(requestRows, ROWS.size()), in2.items());
```
Comment (xtern) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/AsyncSqlCursorImplTest.java:80]: swapped args to (expected, result)

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/AsyncSqlScriptCursorTest.java (hunk 1 above)
Comment (xtern) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/AsyncSqlScriptCursorTest.java]: Done

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/AsyncSqlScriptCursorTest.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        AsyncCursor<Integer> cursor = prepareTestCursor(listener);
+
+        // Ensure request future completes without errors.
+        assertThat(cursor.requestNextAsync(PAGE_SIZE), willCompleteSuccessfully());
+    }
+
+    @Test
+    public void testPrefetchListenerExecutesAsynchronously() throws Exception {
+        CountDownLatch latch = new CountDownLatch(1);
+        TestPageLoadListener listener = new TestPageLoadListener(() -> latch.await(30, TimeUnit.SECONDS));
```
[the remark this answers is not in the document]
Comment (xtern) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/AsyncSqlScriptCursorTest.java]: Done

**Review by lowka — APPROVED** [review]

**Review by korlov42 — CHANGES_REQUESTED** [review]
The goal of the JIRA is to eliminate dependency on the user's input while processing the script. `AsyncSqlCursorImpl#requestNextAsync` is a user input for sure. Besides, the description of the ticket points out to a particular place that triggers prefetch (`AsyncRootNode#prefetch`). End of the prefetch is the moment when `waiting` reaches 0 for the first time. So, we need to trigger listeners when `waiting` reaches 0 after `AsyncRootNode#prefetch` was invoked.

BTW, I think, the name `listener` is more suitable for recurrent events, while for single shot notification 'callback' fits better

**Comment by xtern** [comment]
> The goal of the JIRA is to eliminate dependency on the user's input while processing the script. `AsyncSqlCursorImpl#requestNextAsync` is a user input for sure. Besides, the description of the ticket points out to a particular place that triggers prefetch (`AsyncRootNode#prefetch`). End of the prefetch is the moment when `waiting` reaches 0 for the first time. So, we need to trigger listeners when `waiting` reaches 0 after `AsyncRootNode#prefetch` was invoked.
> 
> BTW, I think, the name `listener` is more suitable for recurrent events, while for single shot notification 'callback' fits better

@korlov42 

Yes :woman_facepalming:, thanks for the detailed explanation. 

I've reworked the patch, please have a look.

**Review by korlov42 — APPROVED** [review]

**Pull request merged**